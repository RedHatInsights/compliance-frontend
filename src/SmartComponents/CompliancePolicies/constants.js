import { gql } from '@apollo/client';

export const QUERY = gql`
  {
    profiles(search: "external = false and canonical = false") {
      edges {
        node {
          id
          name
          description
          refId
          complianceThreshold
          totalHostCount
          osMajorVersion
          policyType
          policy {
            id
            name
          }
          businessObjective {
            id
            title
          }
        }
      }
    }
  }
`;

export const dataMap = {
  id: ['id', 'policy.id'],
  title: ['name', 'policy.name'],
  description: 'description',
  business_objective: 'businessObjective.title',
  compliance_threshold: 'complianceThreshold',
  total_system_count: 'totalHostCount',
  os_major_version: 'osMajorVersion',
  profile_title: 'policyType',
  ref_id: 'refId',
};
