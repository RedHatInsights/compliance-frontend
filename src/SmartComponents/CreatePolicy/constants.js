import gql from 'graphql-tag';

export const BENCHMARKS_QUERY = gql`
  query CP_Benchmarks($filter: String!) {
    benchmarks(search: $filter) {
      nodes {
        id
        latestSupportedOsMinorVersions
        ruleTree
        valueDefinitions {
          defaultValue
          description
          id
          refId
          title
          valueType
        }
        profiles {
          id
          refId
          osMajorVersion
        }
        version
      }
    }
  }
`;

export const BENCHMARKS_RULES_TREES_QUERY = gql`
  query CP_BenchmarksRuleTree($filter: String!) {
    benchmarks(search: $filter) {
      nodes {
        id
        ruleTree
      }
    }
  }
`;

export const BENCHMARKS_VALUE_DEFINITIONS_QUERY = gql`
  query CP_BenchmarksValueDefinitions($filter: String!) {
    benchmarks(search: $filter) {
      nodes {
        id
        latestSupportedOsMinorVersions
        ruleTree
        valueDefinitions {
          defaultValue
          description
          id
          refId
          title
          valueType
        }
        profiles {
          id
          refId
          osMajorVersion
        }
        version
      }
    }
  }
`;

export const PROFILES_QUERY = gql`
  query CP_Profiles($filter: String!) {
    profiles(search: $filter) {
      edges {
        node {
          id
          refId
          osMinorVersion
          benchmark {
            id
            latestSupportedOsMinorVersions
          }
          rules {
            refId
          }
        }
      }
    }
  }
`;
