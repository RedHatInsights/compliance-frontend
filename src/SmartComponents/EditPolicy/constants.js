import { gql } from '@apollo/client';

export const BENCHMARKS_QUERY = gql`
  query EP_Benchmarks($filter: String!) {
    benchmarks(search: $filter) {
      nodes {
        id
        latestSupportedOsMinorVersions
        ruleTree
        valueDefinitions {
          defaultValue
          description
          id
          refId
          title
          valueType
        }
        profiles {
          id
          refId
          ssgVersion
        }
      }
    }
  }
`;

export const MULTIVERSION_QUERY = gql`
  query EP_Profile($policyId: String!) {
    profile(id: $policyId) {
      id
      name
      refId
      external
      description
      totalHostCount
      compliantHostCount
      complianceThreshold
      osMajorVersion
      supportedOsVersions
      lastScanned
      policyType
      policy {
        id
        name
        refId
        profiles {
          id
          parentProfileId
          name
          refId
          osMinorVersion
          osMajorVersion
          values
          benchmark {
            id
            title
            latestSupportedOsMinorVersions
            osMajorVersion
            version
            ruleTree
          }
          rules {
            title
            severity
            rationale
            refId
            description
            remediationAvailable
            identifier
            values
          }
        }
      }
      businessObjective {
        id
        title
      }
      hosts {
        id
        osMinorVersion
        osMajorVersion
      }
    }
  }
`;

export const RULE_VALUE_DEFINITIONS_QUERY = gql`
  query EP_ProfileValueDefinitions($policyId: String!) {
    profile(id: $policyId) {
      id
      policy {
        id
        refId
        profiles {
          id
          parentProfileId
          refId
          benchmark {
            id
            ruleTree
            valueDefinitions {
              defaultValue
              description
              id
              refId
              title
              valueType
            }
          }
        }
      }
    }
  }
`;
