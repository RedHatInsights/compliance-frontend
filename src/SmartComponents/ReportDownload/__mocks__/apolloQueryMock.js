import {
  buildNonCompliantSystems,
  buildUnsupportedSystems,
  buildSystemsWithTestResultProfiles,
} from '@/__factories__/systems';
import { graphqlResult } from '@/__factories__/helpers';

export default async () =>
  graphqlResult(undefined, {
    systems: [
      ...buildSystemsWithTestResultProfiles(1),
      ...buildNonCompliantSystems(1),
      ...buildUnsupportedSystems(1),
    ],
  }).result;
