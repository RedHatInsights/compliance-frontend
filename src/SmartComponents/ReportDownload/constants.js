import { gql } from '@apollo/client';

export const DEFAULT_EXPORT_SETTINGS = {
  compliantSystems: false,
  nonCompliantSystems: true,
  unsupportedSystems: true,
  nonReportingSystems: true,
  topTenFailedRules: true,
  userNotes: undefined,
};

export const GET_SYSTEMS = gql`
  query PDF_Systems(
    $filter: String!
    $policyId: ID
    $perPage: Int
    $page: Int
    $sortBy: [String!]
    $tags: [String!]
  ) {
    systems(
      search: $filter
      limit: $perPage
      offset: $page
      sortBy: $sortBy
      tags: $tags
    ) {
      totalCount
      edges {
        node {
          id
          name
          osMajorVersion
          osMinorVersion
          insightsId
          testResultProfiles(policyId: $policyId) {
            lastScanned
            compliant
            score
            supported
            benchmark {
              version
            }
            rulesFailed
          }
        }
      }
    }
  }
`;

export const GET_PROFILE = gql`
  query PDF_Profile($policyId: String!) {
    profile(id: $policyId) {
      id
      name
      refId
      testResultHostCount
      compliantHostCount
      unsupportedHostCount
      complianceThreshold
      osMajorVersion
      lastScanned
      policyType
      totalHostCount
      policy {
        id
        name
      }
      benchmark {
        id
      }
      businessObjective {
        id
        title
      }
    }
  }
`;

export const GET_RULES = gql`
  query PDF_Profiles($filter: String!, $policyId: ID!) {
    profiles(search: $filter) {
      totalCount
      edges {
        node {
          topFailedRules(policyId: $policyId) {
            refId
            title
            remediationAvailable
            severity
            identifier
            failedCount
          }
        }
      }
    }
  }
`;
