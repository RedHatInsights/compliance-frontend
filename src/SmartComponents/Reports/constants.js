import { gql } from '@apollo/client';

export const QUERY = gql`
  query R_Profiles($filter: String!) {
    profiles(search: $filter, limit: 1000) {
      edges {
        node {
          id
          name
          refId
          description
          policyType
          totalHostCount
          testResultHostCount
          compliantHostCount
          unsupportedHostCount
          osMajorVersion
          complianceThreshold
          businessObjective {
            id
            title
          }
          policy {
            id
            name
          }
          benchmark {
            id
            version
          }
        }
      }
    }
  }
`;
