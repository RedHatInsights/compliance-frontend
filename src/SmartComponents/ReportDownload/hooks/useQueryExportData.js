import { useApolloClient } from '@apollo/client';
import { GET_SYSTEMS, GET_RULES } from '../constants';
import { prepareForExport } from './helpers';
import usePromiseQueue from 'Utilities/hooks/usePromiseQueue';

const useFetchBatched = () => {
  const { isResolving: isLoading, resolve } = usePromiseQueue();

  return {
    isLoading,
    fetchBatched: (fetchFunction, total, filter, batchSize = 50) => {
      const pages = Math.ceil(total / batchSize) || 1;

      const results = resolve(
        [...new Array(pages)].map(
          (_, pageIdx) => () => fetchFunction(batchSize, pageIdx + 1, filter)
        )
      );

      return results;
    },
  };
};

const useSystemsFetch = ({ id: policyId, totalHostCount } = {}) => {
  const client = useApolloClient();
  const { fetchBatched } = useFetchBatched();

  const fetchFunction = (perPage, page) =>
    client.query({
      query: GET_SYSTEMS,
      fetchResults: true,
      fetchPolicy: 'no-cache',
      variables: {
        perPage,
        page,
        filter: `policy_id = ${policyId}`,
        policyId,
      },
    });

  return async () =>
    (await fetchBatched(fetchFunction, totalHostCount)).flatMap(
      ({
        data: {
          systems: { edges },
        },
      }) => edges.map(({ node }) => node)
    );
};

const useFetchRules = ({ id: policyId } = {}) => {
  const client = useApolloClient();

  const fetchFunction = (perPage = 10, page = 1) =>
    client.query({
      query: GET_RULES,
      fetchResults: true,
      fetchPolicy: 'no-cache',
      variables: {
        perPage,
        page,
        filter: `policy_id = ${policyId}`,
        policyId,
      },
    });

  return async () =>
    (await fetchFunction()).data.profiles?.edges.flatMap(
      (edge) => edge.node.topFailedRules
    );
};

// Hook that provides a wrapper function for a preconfigured GraphQL client to fetch export data
const useQueryExportData = (
  exportSettings,
  policy,
  { onComplete, onError } = {
    onComplete: () => undefined,
    onError: () => undefined,
  }
) => {
  const fetchSystems = useSystemsFetch(policy);
  const fetchRules = useFetchRules(policy);

  return async () => {
    try {
      const systems = await fetchSystems();
      const rules = await fetchRules();
      const exportData = prepareForExport(exportSettings, systems, rules);
      onComplete?.(exportData);
      return exportData;
    } catch (error) {
      if (onError) {
        onError?.(error);
        return [];
      } else {
        throw error;
      }
    }
  };
};

export default useQueryExportData;
