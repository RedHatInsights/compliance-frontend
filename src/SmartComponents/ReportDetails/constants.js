import { systemsDataMapper, testResultsDataMapper } from '@/constants';
import { gql } from '@apollo/client';
import dataSerialiser from 'Utilities/dataSerialiser';
import { buildOSObject } from 'Utilities/helpers';
import { apiInstance } from 'Utilities/hooks/useQuery';

export const QUERY = gql`
  query RDWNRS_Profile($policyId: String!) {
    profile(id: $policyId) {
      id
      name
      refId
      totalHostCount
      testResultHostCount
      compliantHostCount
      unsupportedHostCount
      complianceThreshold
      osMajorVersion
      lastScanned
      policyType
      policy {
        id
        name
        profiles {
          benchmark {
            version
          }
        }
      }
      businessObjective {
        id
        title
      }
    }
  }
`;

export const dataMap = {
  id: 'id',
  title: 'name',
  business_objective: 'businessObjective.title',
  compliance_threshold: 'complianceThreshold',
  type: 'policyType',
  os_major_version: 'osMajorVersion',
  profile_title: ['policy.name', 'policyType'],
  ref_id: 'refId',
  assigned_system_count: 'totalHostCount',
  reported_system_count: 'testResultHostCount',
  compliant_system_count: 'compliantHostCount',
  unsupported_system_count: 'unsupportedHostCount',
};

export const processTestResultsData = (data) =>
  dataSerialiser(
    data.map((entry) => ({
      ...entry,
    })),
    testResultsDataMapper
  );

export const processSystemsData = (data) =>
  dataSerialiser(data, systemsDataMapper);

export const fetchReportingCustomOSes = ({ filters, reportId }) =>
  apiInstance.reportTestResultsOS(reportId, null, filters).then(({ data }) => {
    return {
      results: buildOSObject(data),
      total: data?.length || 0,
    };
  });

export const fetchNeverReportedCustomOSes = ({ filters, reportId }) =>
  apiInstance.reportSystemsOS(reportId, null, filters).then(({ data }) => {
    return {
      results: buildOSObject(data),
      total: data?.length || 0,
    };
  });
