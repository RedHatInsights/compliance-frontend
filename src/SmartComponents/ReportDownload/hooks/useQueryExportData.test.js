import { renderHook } from '@testing-library/react';
import { useApolloClient } from '@apollo/client';
import apolloQueryMock from '../__mocks__/apolloQueryMock';
import { DEFAULT_EXPORT_SETTINGS } from '../constants';
import useQueryExportData from './useQueryExportData';

jest.mock('@apollo/client');

describe('useQueryExportData', () => {
  const profile = { id: 'TEST_ID', name: 'TEST_PROFILE', totalHostCount: 3 };

  beforeEach(() => {
    useApolloClient.mockImplementation(() => ({
      query: apolloQueryMock,
    }));
  });

  it('returns a query function', async () => {
    const {
      result: { current: queryExportData },
    } = renderHook(() => useQueryExportData(DEFAULT_EXPORT_SETTINGS, profile));
    expect(typeof queryExportData).toBe('function');
  });

  describe('queryExportData', () => {
    const onComplete = jest.fn();
    const onError = jest.fn();

    afterEach(() => {
      onComplete.mockReset();
      onError.mockReset();
    });

    it('returns data', async () => {
      const {
        result: { current: queryExportData },
      } = renderHook(() =>
        useQueryExportData(DEFAULT_EXPORT_SETTINGS, profile, {
          onComplete,
          onError,
        })
      );
      const result = await queryExportData();

      expect(result.nonCompliantSystemCount).toBe(2);
      expect(onComplete).toHaveBeenCalled();
      expect(onError).not.toHaveBeenCalled();
    });

    it('returns catches errors', async () => {
      useApolloClient.mockImplementation(() => ({
        query: async () => {
          throw 'Error';
        },
      }));
      const {
        result: { current: queryExportData },
      } = renderHook(() =>
        useQueryExportData(DEFAULT_EXPORT_SETTINGS, profile, {
          onComplete,
          onError,
        })
      );

      await queryExportData();

      expect(onComplete).not.toHaveBeenCalled();
      expect(onError).toHaveBeenCalled();
    });
  });
});
