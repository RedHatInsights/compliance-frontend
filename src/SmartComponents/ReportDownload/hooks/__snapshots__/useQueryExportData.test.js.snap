// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useQueryExportData queryExportData returns catches errors 1`] = `Object {}`;

exports[`useQueryExportData queryExportData returns data 1`] = `
Object {
  "compliantSystemCount": undefined,
  "nonCompliantSystemCount": undefined,
  "nonCompliantSystems": 10,
  "topTenFailedRules": 10,
  "unsupportedSystemCount": undefined,
  "unsupportedSystems": 10,
}
`;

exports[`useQueryExportData returns a query function 1`] = `[Function]`;
